### PR TITLE
Add tls alpn config field

### PR DESCRIPTION
**Pull request by cristian-aldea**
This PR aims to add configuration for setting the ALPN advertisement over HTTPS, similar to the `tls-alpn` configuration present in the [jcmoraisjr/haproxy-ingress](https://github.com/jcmoraisjr/haproxy-ingress) implementation.

See: https://haproxy-ingress.github.io/docs/configuration/keys/#tls-alpn

See the message I posted on slack about this: https://haproxy.slack.com/archives/CR9N5UNAZ/p1631543163057500

Here is a brief summary of the commits, but please read the commit messages for details:
- added a default value for the `tls-alpn` annotation
- modify "FrontendEnableSSLOffload" in https.go to use the `tls-alpn` annotation
- update documentation

If anything is missing in my contribution please let me know and I will amend any issues as quickly as I can. I read your contribution guide but I might have missed following any rules specific to this repository.

For linting, I ran `make lint` and saw issues on master and other branches already before I made any contributions, so i can't tell if any changes I did would be blocked due to linting issues.

Let me know if I need to make a PR to a specific version branch (`v1.6`), and not `master`. I saw that other PRs were done with master as a base, so I went with that.

### CODE REVIEW

File: controller/haproxy/api/frontend.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func (c *clientNative) FrontendEdit(frontend models.Frontend) error {
 	return c.nativeAPI.Configuration.EditFrontend(frontend.Name, &frontend, c.activeTransaction, 0)
 }
 
-func (c *clientNative) FrontendEnableSSLOffload(frontendName string, certDir string, alpn bool) (err error) {
+func (c *clientNative) FrontendEnableSSLOffload(frontendName string, certDir string, alpn bool, alpnValue string) (err error) {
```
Comment (Mo3m3n) [inline controller/haproxy/api/frontend.go:54]: I think there is no need to have both `alpn bool` and `alpnValue string`
We can have a single param `alpn string` where `if alpn != ""` then controller sets the alpn field.

File: controller/handler/https.go [hunk 2]
```diff
@@ -159,14 +160,14 @@ func (h HTTPS) Update(k store.K8s, cfg *config.ControllerCfg, api api.HAProxyCli
 	_, errFtSSL := api.FrontendGet(cfg.FrontSSL)
 	if cfg.SSLPassthrough {
 		if errFtSSL != nil {
-			logger.Error(h.enableSSLPassthrough(cfg, api))
+			logger.Error(h.enableSSLPassthrough(k, cfg, api))
```
Comment (Mo3m3n) [inline controller/handler/https.go:163]: Maybe no need to send the whole store but only the tls-alpn value at this stage 
Or you can even have a private field in handler struct that you can set in the main `Update` method:
```
h.alpn = annotations.GetValue("tls-alpn", k.ConfigMaps.Main.Annotations)
```
Then later you can use it when calling `FrontendEnableSSLOffload`
```
api.FrontendEnableSSLOffload(cfg.FrontHTTPS, h.CertDir, h.alpn)
```

**Comment by Mo3m3n** [comment]
Looks good over all, only few remarks about code organization.
It seems the commit subject checker does not accept `FEATURE` we need to fix it cause this tag should be accepted. 
But in the meantime you can simple set the severity `MINOR` instead of `FEATURE/MINOR`

**Comment by Mo3m3n** [comment]
No need for a seperate commit to introduce a default value of an annotation before the annotation is used so you can squash "FEATURE/MINOR: frontend: add tls-alpn default value" in "FEATURE/MEDIUM: https: Implement alpn customization in https frontend"

Same for "MINOR: docs: fix typo for tls-alpn section" you can squash it in "MINOR: docs: include tls-alpn annotation in docs"
And regarding doc commits, the right tag would rather be: "DOC/MINOR"

**Comment by cristian-aldea** [comment]
Thank you for the suggestions, I will implement them tomorrow!

Had a question. Currently we are using a custom image built based on v1.6.7 and including these changes. How long do you think it will take for these changes to be ported to the `v1.6` branch and included in a future release?

**Comment by Mo3m3n** [comment]
We try to avoid backporting new features to stable branch.
But for minor changes like this one, we probably can do it, especially that this one does not change behavior if tls-alpn is not used.
So I think it can be available in next stable release, which can be available before the end of this month

**Comment by cristian-aldea** [comment]
> We try to avoid backporting new features to stable branch.

Just had a quick question about this and your general git workflow. What exactly is your workflow with the master and stable branches? 

For example, say if the feature was bigger, should I have made a PR for the stable branch? What exactly is the master branch used for?

**Comment by oktalz** [comment]
@cristian-aldea 

first some positives :)
 if feature is small and in general does not change behavior of controller and community can benefit from it, we can consider as @Mo3m3n said, to put it in already stable branch like `v1.6`. Quality of life is important :)

In general,
- PRs with new features are mainly for master
- if PR is a bug fix or something that impact current stable version then we back-port it to that stable branch.
- PRs with buxfixes can also be done for specific stable branch, but most of the time master is the branch that should be targeted
- small quality of life features (I would say minor improvements) can be also backported if new release is not planned 'soon'

to put some real labels, here is how it stands now:

branches :
 - `master` - receives new features and fixes
 - `v1.6` - receives only bug fixes (and package updates for security purposes if needed)
 - nightly image is available, but its possibly not as stable as standard images
 
when we decide to release `v1.7.0`:
- a new branch will be created `v1.7` and that branch will behave in same way as now a.6
- after its creation only receive bug fixes, 
- no new feature should appear in 1.7 branch, unless it resolves some internal issues (or some other exception) 
- `v1.6` will still receive bug-fixes for some time 
 
 
  regarding commit messages, we use same system as HAProxy does: `FEATURE/MINOR` should be just `MINOR`. We need to clarify that a bit more on CI

**Comment by cristian-aldea** [comment]
Thank you for the clarifications! Sorry about that Its been a hectic day and I haven't gone around to fixing this PR, it will be done by the end of the day since its very simple changes.

File: controller/handler/https.go (hunk 2 above)
Comment (cristian-aldea) [reply controller/handler/https.go:163]: Yup, that's a lot simpler! I will add a field to the HTTPS struct and set the value at the beginning of the Update method

File: controller/haproxy/api/frontend.go (hunk 1 above)
Comment (cristian-aldea) [reply controller/haproxy/api/frontend.go]: Done, thanks for the suggestion

**Comment by cristian-aldea** [comment]
Hello! Seems like last time one of the e2e checks failed, and I'm not sure why. The error message was rather vague, and one other e2e test timed out.

**Comment by oktalz** [comment]
@cristian-aldea tests unfortunately can fail on Github, its still a matter of figuring out why exactly, 
since locally and in our infrastructure they are passing.
Github it seems also changed few things regarding CI runs, such as CI not starting for first time contributors that needs to be started manually from our side

**Comment by Mo3m3n** [comment]
Yup sequential e2e tests are a flaky on github, am working on this issue.
I have re-run then and all is green.
Thanks @cristian-aldea for the PR it looks good.

**Comment by cristian-aldea** [comment]
> @cristian-aldea tests unfortunately can fail on Github, its still a matter of figuring out why exactly,
> since locally and in our infrastructure they are passing.
> Github it seems also changed few things regarding CI runs, such as CI not starting for first time contributors that needs to be started manually from our side

I'm aware, but there was obviously some flakyness to the tests, since one test failed, another test timed out and another passed despite the fact that my changes shouldn't care about what k8s version is running.

**Comment by cristian-aldea** [comment]
> Yup sequential e2e tests are a flaky on github, am working on this issue.
> I have re-run then and all is green.
> Thanks @cristian-aldea for the PR it looks good.

Thank you for merging the PR! I'm glad i could help. Keep me posted on if this change will be backported to 1.6 soon.

Thank you all for the help, have a nice day.

**Comment by cristian-aldea** [comment]
If it can help with anything, here is the current implementation I made to backport the changes I did to v1.6. This branch is based on the v1.6.7 tag (dc59877c734d1cc748f393363af39b2fbe9da4ef), and has an extra commit 33603f49ad6eeb9545095c3a455028203c4d912c with my backport

https://github.com/cristian-aldea/kubernetes-ingress/tree/custom-build

**Pull request merged**